### PR TITLE
imp: adding 08-wasm build opts with libwasmvm linking disabled

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

Tested adding custom build directive `nolink_libwasmvm` in rollkit app binary compilation.
See: https://github.com/chatton/gm-demo/pull/2 (with successful build discarding libwasmvm dependencies - NOTE: uses fork of wasmvm with build directives)

```
panic: not implemented, please build with cgo enabled or nolink_libwasmvm disabled

goroutine 1 [running]:
github.com/cosmos/ibc-go/modules/light-clients/08-wasm/keeper.NewKeeperWithVM(...)
        github.com/cosmos/ibc-go/modules/light-clients/08-wasm@v0.0.0-20240320102325-afd056184d7e/keeper/keeper_no_vm.go:27
gm/app.RegisterIBC({0x10f524678, 0xc0004a2000})
        gm/app/ibc.go:222 +0x50b
gm/cmd/gmd/cmd.NewRootCmd()
        gm/cmd/gmd/cmd/root.go:110 +0x3d7
main.main()
        gm/cmd/gmd/main.go:14 +0x17
```

When the application binary is compiled removing the calls to `wasmkeeper.NewKeeperWithVM` and `wasmkeeper.NewKeeperWithConfig` the docker image can be run and the entrypoint executes successfully returning the `simd` help text.

closes: #5903 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

## Summary by CodeRabbit

- **New Features**
	- Introduced a Docker build target for WebAssembly simulation applications.
	- Enhanced the Wasm module by integrating WebAssembly VM functionality for better interface and management within Cosmos SDK applications.
	- Added documentation on configuring the Wasm module for different operational modes, including options to disable CGO or selectively link with `libwasmvm`.

- **Refactor**
	- Streamlined Keeper initialization in the Wasm module by consolidating logic into a single struct initialization approach.
	- Removed redundant interface declaration in the Wasm engine code to simplify interface definitions.

- **Chores**
	- Updated the Makefile to include a new build target and improved the `help` target output.

### CODE REVIEW

File: modules/light-clients/08-wasm/internal/ibcwasm/wasm_no_cgo.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/internal/ibcwasm/wasm_no_cgo.go:1]: This file is not strictly necessary so I can remove it.

**Comment by chatton** [comment]
One thing that would be nice to test before merging this is maybe we can try building the gm-demo importing this commit?

File: modules/light-clients/08-wasm/internal/ibcwasm/wasm_no_cgo.go
Comment (chatton) [inline modules/light-clients/08-wasm/internal/ibcwasm/wasm_no_cgo.go:5]: this is basically a no-op right? Do we even need this file?

Comment (damiannolan) [reply modules/light-clients/08-wasm/internal/ibcwasm/wasm_no_cgo.go:5]: yeah I think we can delete it

File: modules/light-clients/08-wasm/internal/ibcwasm/wasm_no_cgo.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/internal/ibcwasm/wasm_no_cgo.go:1]: https://github.com/cosmos/ibc-go/pull/5923/commits/21832c1aca197483091efcb36b8b5268ef0b2276

File: modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/internal/ibcwasm/expected_interfaces.go:13]: This was moved to another file. nt bot

**Comment by damiannolan** [comment]
Let's change this to not use the `cgo` build tags and instead use a separately defined one. Similarly to how we use build tags in `e2e`.

@charleenfei discovered when building binaries that require some `cgo` dependencies for keys etc that some code is not available, and re-enabling the `cgo` build tag would then require the wasmvm dlls to be available. I think we can use a separate tag here to instruct the compiler to use the files when buillding - making it possible to have `CGO_ENABLED=1` but `08-wasm` using types only.

File: modules/light-clients/08-wasm/keeper/keeper_no_vm.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/keeper/keeper_no_vm.go:42]: Agree with code rabbit, we should add some extra info in the godoc to state why this is

Comment (damiannolan) [reply modules/light-clients/08-wasm/keeper/keeper_no_vm.go:42]: I disagree with your suggest - it seems misleading and this module is not used in conjunction with x/wasm.

The purpose of this function is to be present when the binary is compiled with `ibcwasm_novm`, such that it will panic if used by an application developer when wasmvm dependencies are not present on the system.

**Comment by damiannolan** [comment]
I'm going to update this PR to match the build directives used in: https://github.com/CosmWasm/wasmvm/pull/527

**Comment by damiannolan** [comment]
- Tested compilation in docker env using rollkit application in https://github.com/chatton/gm-demo/pull/2 🟢 
- Tested compilation in docker env using go relayer with cgo enabled and `nolink_libwasmvm` (docker image available at `docker.io/damiannolan/rly-local`) 🟢 

Happy to consider this good to go if there is agreement to go ahead with [CosmWasm/wasmvm#527](https://github.com/CosmWasm/wasmvm/pull/527)

File: modules/light-clients/08-wasm/keeper/keeper_no_vm.go
Comment (chatton) [reply modules/light-clients/08-wasm/keeper/keeper_no_vm.go:1]: I think no_vm is more representative of the purpose of these build tags

**Review by chatton — APPROVED** [review]
LGTM

**Comment by damiannolan** [comment]
Will require bump of next wasmvm release. Merging! 🚀

**Pull request merged**
